### ISSUE TITLE
Rollback to last save action on shortcut

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#5001]
**QA**: @farmaazon 

This task is automatically imported from the old Task Issue Board and it was originally created by Adam Obuchowicz.
Original issue is [here](https://www.pivotaltracker.com/story/show/183604295).
<hr />

## Why

**As** Enso IDE user
**I want** to be able to roll back all changes I've done from last named save
**So that** I can discard changes if they turn out to be undesirable

## Acceptance Criteria
```gherkin
Scenario: 
GIVEN I have a project opened and saved with ctrl+s
WHEN I press the shortcut for rolling back
THEN nothing happens
WHEN I create new node, move another one and press the shortcut for rolling back
THEN all changes are reverted.
```

## Blockers:
[#184324365](https://www.pivotaltracker.com/story/show/#184324365) blocked
[#184335600](https://www.pivotaltracker.com/story/show/#184335600) resolved

## Comments:
**Galin Bajlekov** reports a new **STANDUP** for today (2022-12-16):

**Progress:** I've studied the mechanism for loading code from a file and updating the module model and the language server with the new content. It should be finished by 2022-12-22.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll implement a shortcut to trigger the VCS restore operation, and will look into updating the module model with the restored content in a live project. (Enso Bot - Dec 16, 2022)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2022-12-19):

**Progress:** I have a WIP VCS restore and module model update function, but the file to update the module with appears to be read before it's restored from the VCS. It should be finished by 2022-12-22.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll investigate why the file restored from the VCS is not read properly and will look at whether the module updating happens correctly. (Enso Bot - Dec 19, 2022)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2022-12-20):

**Progress:** VCS restore and main module update work now, allowing for a smooth snapshot saving/restoring workflow. There is a mismatch in line endings on windows between the restored VCS snapshots and what the parser expects, which will need to be addressed separately. It should be finished by 2022-12-22.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll look at determining which modules need to be updated besides the main project module, write unit tests for the snapshot restore functionality and open a PR for this task. (Enso Bot - Dec 20, 2022)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2022-12-21):

**Progress:** I've looked further into the issue with changed file line endings when restoring snapshots from the VCS and am in touch with Hubert, awaiting a possible fix. It should be finished by 2022-12-22. (Enso Bot - Dec 21, 2022)
<hr />
**Galin Bajlekov** reports a new **🔴 DELAY** for today (2023-01-09):

**Summary:** There is **21 days** delay in implementation of the [Rollback to last save action on shortcut (#183604295) task](https://www.pivotaltracker.com/story/show/183604295).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Work on this task was on hold due to an issue with line endings in text restored from the VCS. (Enso Bot - Jan 9, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-09):

**Progress:** I've evaluated #4020 implemented by Hubert as a possible fix for the VCS restore issues. That PR also changes a few language server methods, and I've looked into restructuring the code to use those. Unfortunately, the line ending issues still persist. It should be finished by 2023-01-12.

**Next Day:** Next day I will be working on the [#183803361 task](https://www.pivotaltracker.com/story/show/183803361). I'll go through the code and operation one more time and prepare the PR for review. (Enso Bot - Jan 9, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-10):

**Progress:** I've further investigated how the VCS changes affect the reloading of module sources. The current module implementation is designed to only send module change updates to the language server and not the other way around, causing the file versions to become desynchronized. I've started a thread to discuss a possible solution. It should be finished by 2023-01-12.

**Next Day:** Next day I will be working on the [#183803361 task](https://www.pivotaltracker.com/story/show/183803361). I'll go through the code and operation one more time and prepare the PR for review. (Enso Bot - Jan 10, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-11):

**Progress:** Implemented a function to reload a synchronized module from the language server after the server version has been changed due to a VCS restore action. It should be finished by 2023-01-12.

**Next Day:** Next day I will be working on the [#183803361 task](https://www.pivotaltracker.com/story/show/183803361). I'll go through the code and operation one more time and prepare the PR for review. (Enso Bot - Jan 11, 2023)
<hr />
**Galin Bajlekov** reports a new **🔴 DELAY** for today (2023-01-13):

**Summary:** There is **5 days** delay in implementation of the [Rollback to last save action on shortcut (#183604295) task](https://www.pivotaltracker.com/story/show/183604295).
It **will cause 5 days delay** for the delivery of this weekly plan.

**Delay Cause:** The text synchronization between the language server and module now works, but compiling the new source throws errors on the server side that need to be investigated. (Enso Bot - Jan 13, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-13):

**Progress:** Text synchronization between the language server and module content is now correctly performed on VCS reload. However, after reloading no results or previews are updated. It should be finished by 2023-01-17.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll continue investigating what causes the issues in computing results after a VCS reload. (Enso Bot - Jan 13, 2023)
<hr />
**Galin Bajlekov** reports a new **🔴 DELAY** for today (2023-01-18):

**Summary:** There is **3 days** delay in implementation of the [Rollback to last save action on shortcut (#183604295) task](https://www.pivotaltracker.com/story/show/183604295).
It **will cause 1 days delay** for the delivery of this weekly plan.

**Delay Cause:** Delay due to waiting for bug fixes regarding compiling of the reloaded module content. (Enso Bot - Jan 18, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-18):

**Progress:** I've looked at implementing file updating based on the `FileEdit`s provided by the `text/didChange` notification instead of reloading the whole file on VCS restore. It should be finished by 2023-01-20.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll continue with applying the `FileEdit`s, and finalize the PR for the VCS restore functionality. (Enso Bot - Jan 18, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-19):

**Progress:** I've implemented the VCS restore based on the file edits provided by the `text/didChange` notification. I spent some time finding a bug in the range of changes to apply as reported by `text/didChange`, which Hubert swiftly fixed. It should be finished by 2023-01-20.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll clean up the code and test the VCS restore functionality more thoroughly, and prepare the PR for review. (Enso Bot - Jan 19, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-20):

**Progress:** Spent more time trying to track down issues with compiling reloaded content when using the `FileEdit`s from the `text/didChange` notification. It should be finished by 2023-01-20.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll try to provide more information about what's happening on my end with the communication with the language server on VCS rollback that might be causing these issues. In the meantime, I'll clean up the branch to be ready for review when the issues are fixed. (Enso Bot - Jan 20, 2023)
<hr />
**Galin Bajlekov** reports a new **🔴 DELAY** for today (2023-01-23):

**Summary:** There is **5 days** delay in implementation of the [Rollback to last save action on shortcut (#183604295) task](https://www.pivotaltracker.com/story/show/183604295).
It **will cause 5 days delay** for the delivery of this weekly plan.

**Delay Cause:** Delay due to waiting for bug fixes regarding compiling of the reloaded module content. I've provided Hubert with more info, and have scheduled a call with Hubert, Adam and Wojciech to discuss what the best approach is to wrap this up as soon as possible. (Enso Bot - Jan 23, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-23):

**Progress:** I've looked at the communication with the language server when processing a `text/didChange` notification to see if there are differences compared to the working solution of reloading the whole file content. In the meantime I've done a few QA reviews and addressed comments to the component browser namespace categories PR. It should be finished by 2023-01-25.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll clean up this PR to be ready for review if the issues with applying the edits from the `text/didChange` notifications get fixed, or alternatively if we decide to revert to reloading the whole file on change. (Enso Bot - Jan 23, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-24):

**Progress:** I've cleaned up the code for the VCS restore, and investigated whether the computed data is preserved on VCS restore. It should be finished by 2023-01-25.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll check if there's more I can do regarding the cache invalidating issue, or one of the remaining VCS issues. If not, I'll start on the next task on my timeline. (Enso Bot - Jan 24, 2023)
<hr />
**Galin Bajlekov** reports a new **🔴 DELAY** for today (2023-01-27):

**Summary:** There is **5 days** delay in implementation of the [Rollback to last save action on shortcut (#183604295) task](https://www.pivotaltracker.com/story/show/183604295).
It **will cause 1 days delay** for the delivery of this weekly plan.

**Delay Cause:** The delay is due to being blocked by issues on the engine side. (Enso Bot - Jan 27, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-01-27):

**Progress:** I've looked into the caching of computed node data on VCS restore. The cache was unnecessarily invalidated on restore, as only metadata was altered when parsing the restored file content. Changing that to invalidate only the changed metadata showed that the engine does not re-evaluate changed nodes on reload. A task was created for that and Dmitry will look into it. It should be finished by 2023-01-30.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll continue testing the VCS restore functionality, go through my code one more time, and prepare the PR for review. (Enso Bot - Jan 27, 2023)
<hr />
**Galin Bajlekov** reports a new **🔴 DELAY** for today (2023-02-01):

**Summary:** There is **4 days** delay in implementation of the [Rollback to last save action on shortcut (#183604295) task](https://www.pivotaltracker.com/story/show/183604295).
It **will cause 2 days delay** for the delivery of this weekly plan.

**Delay Cause:** The delay is due to cleaning up the code and writing tests taking longer than expected. (Enso Bot - Feb 1, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-02-01):

**Progress:** I've cleaned up the code to apply and synchronize text edits received from a VCS restore. I've been wrapping my head around the tests to exercise the new synchronization logic and will finish the test tomorrow. It should be finished by 2023-02-03.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll finish unit tests for the implemented logic, and clean up the PR for review. (Enso Bot - Feb 1, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-02-02):

**Progress:** I've been working on unit tests for the file content synchronization on VCS restore, adapting the structure to be able to change the file content on the side of the mocked language server. This went slower than expected due to the long compilation time of the WASM tests and is still in progress. It should be finished by 2023-02-03.

**Next Day:** Next day I will be working on the [#183604295 task](https://www.pivotaltracker.com/story/show/183604295). I'll finish unit tests for the implemented logic, and clean up the PR for review. (Enso Bot - Feb 2, 2023)
<hr />
**Galin Bajlekov** reports a new **STANDUP** for today (2023-02-03):

**Progress:** I finished implementing the unit test for file content changing on the language server side. I've cleaned up the code and have opened the PR for review. It should be finished by 2023-02-03.

**Next Day:** Next day I will be working on the [#184215360 task](https://www.pivotaltracker.com/story/show/184215360). I'll continue working on the debug toggle for displaying private entries in the component browser. (Enso Bot - Feb 3, 2023)
<hr />
